include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -0,0 +1,286 @@
+);
+#undef query
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+}
+
+template<typename T, bool IsAnisotropic=false NBL_STRUCT_CONSTRAINABLE>
+struct GGX;
+
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<T>)
+struct GGX<T,false NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
+{
+    using scalar_type = T;
+
+    // trowbridge-reitz
+    template<class MicrofacetCache NBL_FUNC_REQUIRES(ReadableIsotropicMicrofacetCache<MicrofacetCache>)
+    scalar_type D(NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        scalar_type denom = scalar_type(1.0) - one_minus_a2 * cache.getNdotH2();
+        return a2 * numbers::inv_pi<scalar_type> / (denom * denom);
+    }
+
+    template<class Query NBL_FUNC_REQUIRES(ggx_concepts::DG1BrdfQuery<Query>)
+    scalar_type DG1(NBL_CONST_REF_ARG(Query) query)
+    {
+        return scalar_type(0.5) * query.getNdf() * query.getG1over2NdotV();
+    }
+
+    template<class Query, class MicrofacetCache NBL_FUNC_REQUIRES(ggx_concepts::DG1BsdfQuery<Query> && ReadableIsotropicMicrofacetCache<MicrofacetCache>)
+    scalar_type DG1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        scalar_type NG = query.getNdf() * query.getG1over2NdotV();
+        scalar_type factor = scalar_type(0.5);
+        if (cache.isTransmission())
+        {
+            const scalar_type VdotH_etaLdotH = (cache.getVdotH() + query.getOrientedEta() * cache.getLdotH());
+            // VdotHLdotH is negative under transmission, so this factor is negative
+            factor *= -scalar_type(2.0) * cache.getVdotHLdotH() / (VdotH_etaLdotH * VdotH_etaLdotH);
+        }
+        return NG * factor;
+    }
+
+    scalar_type devsh_part(scalar_type NdotX2)
+    {
+        assert(a2 >= numeric_limits<scalar_type>::min);
+        return sqrt(a2 + one_minus_a2 * NdotX2);
+    }
+
+    scalar_type G1_wo_numerator(scalar_type absNdotX, scalar_type NdotX2)
+    {
+        return scalar_type(1.0) / (absNdotX + devsh_part(NdotX2));
+    }
+
+    scalar_type G1_wo_numerator_devsh_part(scalar_type absNdotX, scalar_type devsh_part)
+    {
+        // numerator is 2 * NdotX
+        return scalar_type(1.0) / (absNdotX + devsh_part);
+    }
+
+    // without numerator, numerator is 2 * NdotV * NdotL, we factor out 4 * NdotV * NdotL, hence 0.5
+    template<class Query, class LS, class Interaction NBL_FUNC_REQUIRES(ggx_concepts::G2XQuery<Query> && LightSample<LS> && surface_interactions::Isotropic<Interaction>)
+    scalar_type correlated(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction)
+    {
+        BxDFClampMode _clamp = query.getClampMode();
+        assert(_clamp != BxDFClampMode::BCM_NONE);
+
+        scalar_type Vterm = _sample.getNdotL(_clamp) * query.getDevshV();
+        scalar_type Lterm = interaction.getNdotV(_clamp) * query.getDevshL();
+        return scalar_type(0.5) / (Vterm + Lterm);
+    }
```

Comment:
id you want to return without numerator make a function called `correlated_wo_numerator` if we break the convention it becomes really confusing from a documentation/concept side of things